pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go

```diff
@@ -0,0 +1,616 @@
+	return quotaManager
+}
+
+func (gqm *GroupQuotaManager) SetScaleMinQuotaEnabled(flag bool) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.scaleMinQuotaEnabled = flag
+	klog.V(3).Infof("Set ScaleMinQuotaEnabled, flag:%v", gqm.scaleMinQuotaEnabled)
+}
+
+func (gqm *GroupQuotaManager) UpdateClusterTotalResource(deltaRes v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	klog.V(3).Infof("UpdateClusterResource deltaRes:%v", deltaRes)
+	gqm.updateClusterTotalResourceNoLock(deltaRes)
+}
+
+func (gqm *GroupQuotaManager) updateClusterTotalResourceNoLock(deltaRes v1.ResourceList) {
+	gqm.totalResource = quotav1.Add(gqm.totalResource, deltaRes)
+
+	sysAndDefaultUsed := quotav1.Add(gqm.quotaInfoMap[extension.DefaultQuotaName].GetUsed(), gqm.quotaInfoMap[extension.SystemQuotaName].GetUsed())
+	totalResNoSysOrDefault := quotav1.Subtract(gqm.totalResource, sysAndDefaultUsed)
+
+	diffRes := quotav1.Subtract(totalResNoSysOrDefault, gqm.totalResourceIndeed)
+
+	if !quotav1.IsZero(diffRes) {
+		gqm.totalResourceIndeed = totalResNoSysOrDefault.DeepCopy()
+		gqm.quotaTreeWrapperMap[extension.RootQuotaName].SetClusterTotalResource(totalResNoSysOrDefault)
+		klog.V(3).Infof("UpdateClusterResource finish totalResourceIndeed:%v", gqm.totalResourceIndeed)
+	}
+}
+
+func (gqm *GroupQuotaManager) GetClusterTotalResource() v1.ResourceList {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	return gqm.totalResource.DeepCopy()
+}
+
+func (gqm *GroupQuotaManager) UpdateGroupDeltaRequest(quotaName string, deltaReq v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.updateGroupDeltaRequestNoLock(quotaName, deltaReq)
+}
+
+// updateGroupDeltaRequestNoLock no need lock gqm.lock
+func (gqm *GroupQuotaManager) updateGroupDeltaRequestNoLock(quotaName string, deltaReq v1.ResourceList) {
+	curToAllParInfos := gqm.getCurToAllParentGroupQuotaInfoNoLock(quotaName)
+	allQuotaInfoLen := len(curToAllParInfos)
+	if allQuotaInfoLen <= 0 {
+		return
+	}
+
+	defer gqm.scopedLockForQuotaInfo(curToAllParInfos)()
+
+	gqm.updateGroupDeltaRequestTopoRecursiveNoLock(deltaReq, curToAllParInfos)
+}
+
+// updateGroupDeltaRequestTopoRecursiveNoLock update the quota of a node, also need update all parentNode, the lock operation
+// of all quotaInfo is done by gqm. scopedLockForQuotaInfo, so just get treeWrappers' lock when calling treeWrappers' function
+func (gqm *GroupQuotaManager) updateGroupDeltaRequestTopoRecursiveNoLock(deltaReq v1.ResourceList, curToAllParInfos []*QuotaInfo) {
+	quotaInfo := curToAllParInfos[0]
+	oldSubLimitReq := quotaInfo.getLimitRequestNoLock()
+	quotaInfo.addRequestNonNegativeNoLock(deltaReq)
+	newSubLimitReq := quotaInfo.getLimitRequestNoLock()
+	treeWrapperPtr := gqm.getTreeWrapperByNameNoLock(quotaInfo.ParentName)
+	if treeWrapperPtr == nil {
+		klog.Errorf("treeWrapper not exist! quotaInfoName:%v parentName:%s", quotaInfo.Name, quotaInfo.ParentName)
+		return
+	}
+	if treeWrapperPtr.NeedUpdateOneGroupRequest(quotaInfo) {
```

Comment:
the process seems wrong here, suppose parent has two leaf node,  if one leaf node update request, why you directly use one leaf node's limit request to update the parent whole limit request? the parent request should be always its sub leaf nodes request total.